src/home/homework-blob.c

```diff
@@ -0,0 +1,242 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "copy.h"
+#include "fileio.h"
+#include "fd-util.h"
+#include "format-util.h"
+#include "fs-util.h"
+#include "home-util.h"
+#include "homework-blob.h"
+#include "homework.h"
+#include "install-file.h"
+#include "macro.h"
+#include "path-util.h"
+#include "recurse-dir.h"
+#include "rm-rf.h"
+#include "sha256.h"
+#include "string-util.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+#include "utf8.h"
+
+static int copy_one_blob(
+                int src_fd,
+                int dest_dfd,
+                const char *name,
+                uint64_t *total_size,
+                uid_t uid,
+                Hashmap *manifest) {
+        _cleanup_(unlink_and_freep) char *dest_tmpname = NULL;
+        _cleanup_close_ int dest = -EBADF;
+        uint8_t hash[SHA256_DIGEST_SIZE], *known_hash;
+        off_t initial, size;
+        int r;
+
+        assert(src_fd >= 0);
+        assert(dest_dfd >= 0);
+        assert(name);
+        assert(total_size);
+        assert(uid_is_valid(uid));
+        assert(manifest);
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("Blob %s has invalid filename. Skipping.", name);
+                return 0;
+        }
+
+        known_hash = hashmap_get(manifest, name);
+        if (!known_hash) {
+                log_warning("Blob %s is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (fd_verify_regular(src_fd) < 0) {
+                log_warning("Blob %s is not a regular file. Skipping.", name);
```

Comment:
and then pass it in `log_warning_errno(r, "…`

as a general rule: if we have an error code, we pass it to the log functions if we log. even if we don't use %m. That's because these log functions typically generate structured log messages, and there it makes sense to see the original error.